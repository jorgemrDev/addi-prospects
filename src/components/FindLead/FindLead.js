import React, { useState } from "react";
import { getLeadInformation } from "../../services/crm/crmService";
import ProspectsForm from "../ProspectsForm";

const emptyLead = {
  idNationalNumber: "",
};
const STATUS = {
  IDLE: "IDLE",
  SUBMITTED: "SUBMITTED",
  SUBMITTING: "SUBMITTING",
  COMPLETED: "COMPLETED",
};

export default function FindLead() {
  const [prospect, setProspect] = useState(emptyLead);
  const [status, setStatus] = useState(STATUS.IDLE);
  const [foundLead, setFoundLead] = useState(false);
  const [leadInformation, setLeadInformation] = useState(emptyLead);
  const errors = getErrors(prospect);
  const isValid = Object.keys(errors).length === 0;

  async function handleSubmit(event) {
    event.preventDefault();
    setStatus(STATUS.SUBMITTING);
    setFoundLead(false);
    setLeadInformation(emptyLead);
    if (isValid) {
      let lead = await getLeadInformation(parseInt(prospect.idNationalNumber));
      if (lead && lead[0]) {
        setFoundLead(true);
        setLeadInformation(lead[0]);
      }
      setStatus(STATUS.COMPLETED);
    } else {
      setStatus(STATUS.SUBMITTED);
    }
  }

  function handleChange(e) {
    e.persist(); // persist the event
    setProspect((curLeadIdentification) => {
      return {
        ...curLeadIdentification,
        [e.target.id]: e.target.value,
      };
    });
  }

  function getErrors(prospect) {
    const result = {};
    if (!prospect.idNationalNumber)
      result.idNationalNumber = "National Identification is required";
    return result;
  }

  function clear() {
    setStatus(STATUS.IDLE);
    setFoundLead(false);
    setProspect(emptyLead);
  }

  if (foundLead) {
    return (
      <ProspectsForm prospect={leadInformation} clear={clear}></ProspectsForm>
    );
  }

  return (
    <>
      <h1>Lead Info</h1>
      <div class="container">
        {!isValid && status === STATUS.SUBMITTED && (
          <div role="alert">
            <p>Please fix the following errors:</p>
            <ul>
              {Object.keys(errors).map((key) => {
                return <li key={key}>{errors[key]}</li>;
              })}
            </ul>
          </div>
        )}

        <form onSubmit={handleSubmit}>
          <div>
            <div>
              <label htmlFor="idNationalNumber">Id National Number</label>
              <br />
              <input
                id="idNationalNumber"
                type="text"
                value={prospect.idNationalNumber}
                onChange={handleChange}
              />
            </div>
          </div>
          <div>
            <input type="submit" className="btn btn-primary" value="Find" />
          </div>
        </form>
        {status === STATUS.COMPLETED && !foundLead && (
          <div role="alert">
            <p>Document not found</p>
          </div>
        )}
      </div>
    </>
  );
}
