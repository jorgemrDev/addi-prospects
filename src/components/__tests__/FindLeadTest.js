import React from "react";
import FindLead from "../FindLead";
import renderer from "react-test-renderer";

test("Render Loading correctly", () => {
  const component = renderer.create(<FindLead></FindLead>);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
