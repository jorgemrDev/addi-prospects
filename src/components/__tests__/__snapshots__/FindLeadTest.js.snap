// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Loading correctly 1`] = `
Array [
  <h1>
    Lead Info
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <div>
      <div>
        <label
          htmlFor="idNationalNumber"
        >
          Id National Number
        </label>
        <br />
        <input
          id="idNationalNumber"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
    <div>
      <input
        className="btn btn-primary"
        type="submit"
        value="Find"
      />
    </div>
  </form>,
]
`;
